src/etc/inc/plugins.inc.d/captiveportal.inc

```diff
@@ -0,0 +1,190 @@
+    if ($cp->isEnabled()) {
+        $services[] = array(
+            'pidfile' => '/var/run/lighttpd-api-dispatcher.pid',
+            'description' => gettext('Captive Portal'),
+            'configd' => array(
+                'restart' => array('captiveportal restart'),
+                'start' => array('captiveportal start'),
+                'stop' => array('captiveportal stop'),
+            ),
+            'name' => 'captiveportal',
+        );
+    }
+
+    return $services;
+}
+
+function captiveportal_cron()
+{
+    global $config;
+
+    $jobs = [];
+
+    if (!empty($config['system']['captiveportalbackup']) && $config['system']['captiveportalbackup'] > 0) {
+        $jobs[]['autocron'] = array(
+            '/usr/local/etc/rc.syshook.d/backup/20-captiveportal',
+            '0',
+            '*/' . $config['system']['captiveportalbackup']
+        );
+    }
+
+    return $jobs;
+}
+
+function captiveportal_syslog()
+{
+    $logfacilities = [];
+
+    $logfacilities['portalauth'] = ['facility' => ['captiveportal']];
+
+    return $logfacilities;
+}
+
+function captiveportal_firewall($fw)
+{
+    global $config;
+
+    $cp = new \OPNsense\CaptivePortal\CaptivePortal();
+    if ($cp->isEnabled()) {
+        foreach ($cp->zones->zone->iterateItems() as $zone) {
+            if (empty((string)$zone->enabled)) {
+                continue;
+            }
+
+            $zoneid = (string)$zone->zoneid;
+            $uuid = $zone->getAttribute('uuid');
+            // register anchor, to be filled with ether rules for accounting
+            $fw->registerAnchor("captiveportal_zone_{$zoneid}", "ether", 0, "head", false, (string)$zone->interfaces);
+
+            foreach (explode(',', $zone->interfaces) as $intf) {
+                // allow DNS
+                $fw->registerFilterRule(
+                    1,
+                    [
+                        'type' => 'pass',
+                        'interface' => $intf,
+                        'protocol' => 'tcp/udp',
+                        'direction' => 'in',
+                        'from' => 'any',
+                        'to' => '(self)',
+                        'to_port' => 53,
+                        'descr' => "Allow DNS for Captive Portal (zone {$zoneid})",
+                        'log' => !isset($config['syslog']['nologdefaultpass']),
+                        '#ref' => "ui/captiveportal#edit={$uuid}",
+                    ]
+                );
+
+                foreach (['80', '443'] as $to_port) {
+                    $rdr_port = $to_port === '443' ? (8000 + (int)$zoneid) : (9000 + (int)$zoneid);
+
+                    // XXX the forwarding rules can be merged if
+                    // source inversion is supported by registerForwardRule
+                    // for now, skip rdr on the relevant table
+
+                    // disable rdr for authenticated clients, jump to filter pass rule
+                    $fw->registerForwardRule(
```

Comment:
doesn't the `_not` suffix already work? (e.g. from_not)
https://github.com/opnsense/core/blob/03f96eb008eae0de6ba848511bfd6e3f4edabe47/src/opnsense/mvc/app/library/OPNsense/Firewall/Rule.php#L98-L100

all rules types use `reader()` to handle generics.

Reply:
@AdSchellevis Just tested it, seems `from_not` isn't quite there yet for both filter and forwarding rules.

```
'from_not' => '192.168.1.1/32' --> results in "any".
'from_not' => "__captiveportal_zone_{$zoneid}" --> skips parsing for fwd rule, "any" for filter rule
'from_not' => "<__captiveportal_zone_{$zoneid}>" --> same as above
```

The following does work for both filter and forward rules:
```
'source' => ['not' => true, 'address' => "<__captiveportal_zone_{$zoneid}>"]
```

Reply:
@swhite2 `from_not` is a boolean, so it's `from` => `192.168.1.1/32` + `from_not` >=1